[3D] Point cloud editing paintbrush tool

Withalion opened the pull request:
## Description

This PR adds new tool for editing point clouds. The new tool supports free movement, size change, highlighting of already selected area. Users can change the size of selection area by scrolling and the tool starts selecting on left click + drag. Cursor also shows loading while the point cloud is edited.


https://github.com/user-attachments/assets/eac431fa-eb5e-4e77-a46d-fe0632cb828e

Funded by: Klimadatastyrelsen, Septima
<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

Withalion committed:
Add PaintBrush tool for point cloud editing WIP

Withalion committed:
Fix camera movement & paint brush tool appearance

# Conflicts:
#	src/3d/qgsrubberband3d.cpp
#	src/3d/qgsrubberband3d.h

Withalion committed:
Move paint brush tool icon, refactor a bit of code

Withalion committed:
Add editing algo to paint brush tool & selected area highlighter

Withalion committed:
Improve PaintBrush highlighting & add polygon support in RubberBand3D

Withalion committed:
Use undostack in paintbrush tool & refactor deps

Withalion commented:
This PR builds on work done in #60378

saberraz commented:
I have tried this tool. Very nice addition. Some usability issues:
- Changing paintbrush size is controlled by the middle mouse key (wheel), but rolling back and forth. This is always dedicated for map zooming in/out and it is confusing. It is best if you combine it with a hot-key.
- More the above point, it would be good to have a dedicated pixel size for the radius of the paintbrush. 
- While using the tool, I pressed the Escape key and it crashed QGIS.

Withalion committed:
Add more controls to paintbrush tool

Withalion commented:
Paintbrush tool now supports cancellation of selection while still dragging by pressing `ESC` and the size of paintbrush changes by `CTRL`  + mouse wheel.

Withalion committed:
Add movement state for paintbrush tool

Withalion committed:
Refactor includes

Withalion committed:
Fix rebase issues on upgraded RubberBand3D

Withalion committed:
Rework point cloud 3D editing tools

Qgs3DMapToolPointCloudChangeAttribute becomes new base class for polygon
tool and paintbrush tool. Processing method has been moved to new
Qgs3DEditUtils class.

Withalion committed:
Fix crash on 3D window close

Crash occurs when edit tool is active and 3D window is closed

wonder-sk commented on src/app/3d/qgs3deditutils.h:
```diff
@@ -0,0 +1,81 @@
+};
+
+/**
+ * \ingroup 3d
```
we're in "app" here - "3d" is for qgis_3d library

wonder-sk commented on src/app/3d/qgs3deditutils.h:
```diff
@@ -0,0 +1,81 @@
+/**
+ * \ingroup 3d
+ * \brief Utility functions used by 3D editing tools.
+ * \note Not available in Python bindings
```
not needed to state this, because nothing in src/app gets python bindings...

wonder-sk commented on src/app/3d/qgs3deditutils.h:
```diff
@@ -0,0 +1,81 @@
+ * \note Not available in Python bindings
+ * \since QGIS 3.44
+ */
+class Qgs3DEditUtils
```
I would suggest to rename this class to make it clear it is for point cloud editing

wonder-sk commented on src/app/3d/qgs3deditutils.h:
```diff
@@ -0,0 +1,81 @@
+#ifndef QGS3DEDITUTILS_H
+#define QGS3DEDITUTILS_H
+
+#define SIP_NO_FILE
```
no need for SIP_NO_FILE given we're in src/app

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -98,11 +95,16 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
+  QToolButton *editingToolsButton = qobject_cast<QToolButton *>( mEditingToolBar->widgetForAction( mEditingToolsAction ) );
+  editingToolsButton->setPopupMode( QToolButton::ToolButtonPopupMode::InstantPopup );
+  QAction *actionPointCloudChangeAttributeTool = mEditingToolsMenu->addAction( QIcon( QgsApplication::iconPath( QStringLiteral( "mActionSelectPolygon.svg" ) ) ), tr( "Polygon selector" ), this, &Qgs3DMapCanvasWidget::polygonTool );
+  QAction *actionPaintBrush = mEditingToolsMenu->addAction( QIcon( QgsApplication::iconPath( QStringLiteral( "propertyicons/rendering.svg" ) ) ), tr( "Paint Brush Selector" ), this, &Qgs3DMapCanvasWidget::paintBrush );
```
I think it is one word - "Paintbrush"

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -98,11 +95,16 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
+  mEditingToolBar->addAction( mEditingToolsAction );
+  QToolButton *editingToolsButton = qobject_cast<QToolButton *>( mEditingToolBar->widgetForAction( mEditingToolsAction ) );
+  editingToolsButton->setPopupMode( QToolButton::ToolButtonPopupMode::InstantPopup );
+  QAction *actionPointCloudChangeAttributeTool = mEditingToolsMenu->addAction( QIcon( QgsApplication::iconPath( QStringLiteral( "mActionSelectPolygon.svg" ) ) ), tr( "Polygon selector" ), this, &Qgs3DMapCanvasWidget::polygonTool );
```
Use title case - "Polygon Selector"

But maybe to reuse terms used for map selection tool in 2D canvas - how about "Select by Polygon", and "Select by Paintbrush" ?

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -161,6 +163,7 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
   actionGroup->addAction( actionCameraControl );
   actionGroup->addAction( actionIdentify );
   actionGroup->addAction( actionMeasurementTool );
+  actionGroup->addAction( actionPaintBrush );
```
```suggestion
  actionGroup->addAction( actionPaintbrush );
```

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -463,13 +471,30 @@ void Qgs3DMapCanvasWidget::measureLine()
 }
 
-void Qgs3DMapCanvasWidget::changePointCloudAttribute()
+void Qgs3DMapCanvasWidget::paintBrush()
```
How about this? it is a bit long, but this is a class not specific to point clouds...
```suggestion
void Qgs3DMapCanvasWidget::changePointCloudAttributeByPaintbrush()
```

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -463,13 +471,30 @@ void Qgs3DMapCanvasWidget::measureLine()
     return;
 
-  mCanvas->setMapTool( action->isChecked() ? mMapToolPointCloudChangeAttribute : nullptr );
+  mCanvas->setMapTool( nullptr );
```
is it necessary to first reset the map tool to nullptr? it should not be :slightly_smiling_face:

Withalion replied:
this fixed a bug where QGIS was crashing

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -463,13 +471,30 @@ void Qgs3DMapCanvasWidget::measureLine()
+  mEditingToolsAction->setIcon( action->icon() );
+}
+
+void Qgs3DMapCanvasWidget::polygonTool()
```
also rename this please - it should be clear it is for point cloud editing

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.h:
```diff
@@ -69,20 +69,20 @@ class APP_EXPORT Qgs3DMapCanvasWidget : public QWidget
     void setMainCanvas( QgsMapCanvas *canvas );
 
-    Qgs3DMapCanvas *mapCanvas3D() { return mCanvas; }
+    Qgs3DMapCanvas *mapCanvas3D() const { return mCanvas; }
```
why are we changing these functions to `const` ? it can be a bit deceptive - it says it is a const function, but it returns a non-const pointer to internal objects... I would prefer to keep this one and others below non-const

Withalion replied:
As I understand it, `const`  functions specify that the function doesn't change the underlying variables, not that it returns `const` pointer

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.h:
```diff
@@ -140,20 +141,21 @@ class APP_EXPORT Qgs3DMapCanvasWidget : public QWidget
     Qgs3DMapToolIdentify *mMapToolIdentify = nullptr;
     Qgs3DMapToolMeasureLine *mMapToolMeasureLine = nullptr;
-    Qgs3DMapToolPointCloudChangeAttribute *mMapToolPointCloudChangeAttribute = nullptr;
+    std::unique_ptr<Qgs3DMapToolPointCloudChangeAttribute> mMapToolChangeAttribute;
```
Qgs3DMapTool is QObject-based, and unique_ptr is not the best fit unfortunately... because QObject-based objects can be deleted by their parents. There is `QObjectUniquePtr` defined in qgis_core library that should work as unique_ptr, and be aware of QObject as well

wonder-sk commented on src/app/3d/qgs3dmaptoolpolygon.h:
```diff
@@ -0,0 +1,56 @@
+class QgsRubberBand3D;
+class QgsPointXY;
+
+class Qgs3DMapToolPolygon : public Qgs3DMapToolPointCloudChangeAttribute
```
please rename the class to make it clear it is related to point cloud editing...

wonder-sk commented on src/app/3d/qgs3dmaptoolpointcloudchangeattribute.h:
```diff
@@ -66,32 +33,19 @@ class Qgs3DMapToolPointCloudChangeAttribute : public Qgs3DMapTool
+    //! Calculate selection and edit set attribute to new value
+    virtual void run();
+    //! Clear selection
+    virtual void restart();
```
do run() and restart() need to be virtual functions? they do not seem to be called from the base class - so they could be only defined as normal functions in subclasses

wonder-sk commented on src/app/3d/qgs3dmaptoolpointcloudchangeattribute.h:
```diff
@@ -17,47 +17,14 @@
+/**
+ * \ingroup App
+ * \brief Base class for map tools editing point clouds on 3D map canvas.
+ * \note Not available in Python bindings
+ * \note Reworked in QGIS 3.44 to be virtual class for editing tools
```
this could be removed

wonder-sk commented on src/app/3d/qgs3dmaptoolpaintbrush.h:
```diff
@@ -0,0 +1,64 @@
+#include "qgs3dmaptoolpointcloudchangeattribute.h"
+
+class QgsRubberBand3D;
+class Qgs3DMapToolPaintBrush : public Qgs3DMapToolPointCloudChangeAttribute
```
rename the class please to make it obvious it is for point cloud editing

wonder-sk commented on src/app/3d/qgs3dmaptoolpaintbrush.cpp:
```diff
@@ -0,0 +1,165 @@
+  QgsTemporaryCursorOverride busyCursor( Qt::WaitCursor );
+
+  const QgsGeometry searchSkeleton = QgsGeometry( new QgsLineString( mDragPositions ) );
+  const QgsGeometry searchPolygon = searchSkeleton.buffer( mSelectionRubberBand->width() / 2, 37 );
```
why 37 vertices per quarter of a circle? :slightly_smiling_face:   It feels like too much - generally we want fewer vertices so that the intersection test does not take too long. Probably 4-8 should be enough

wonder-sk commented on src/app/3d/qgs3dmaptoolpaintbrush.cpp:
```diff
@@ -0,0 +1,165 @@
+void Qgs3DMapToolPaintBrush::generateHighlightArea()
+{
+  const QgsGeometry searchSkeleton = QgsGeometry( new QgsLineString( mDragPositions ) );
+  const QgsGeometry searchGeometry = searchSkeleton.buffer( mSelectionRubberBand->width() / 2, 37 );
```
same here regarding number of segments per circle quarter

wonder-sk commented on src/app/3d/qgs3dmaptoolpaintbrush.cpp:
```diff
@@ -0,0 +1,165 @@
+  const QgsSettings settings;
+  const bool reverseZoom = settings.value( QStringLiteral( "qgis/reverse_wheel_zoom" ), false ).toBool();
+  const bool shrink = reverseZoom ? event->angleDelta().y() < 0 : event->angleDelta().y() > 0;
+  double zoomFactor = shrink ? 0.75 : 1.5;
+  // "Normal" mouse have an angle delta of 120, precision mouses provide data faster, in smaller steps
+  zoomFactor = 1.0 + ( zoomFactor - 1.0 ) / 120.0 * std::fabs( event->angleDelta().y() );
```
let's simplify this - e.g.:
```
double zoomFactor = ( shrink ? 0.75 : 1.5 ) / 120.0 * std::fabs( event->angleDelta().y() );
```
(if I got the math right)

By the way - it seems like the zooming in and zooming out does not have the same speed - when shrinking, the change is 33% (1.0 / 0.75) and when expanding, the change is 50% (1.5 / 1.0)   ... but maybe I got the math wrong...

wonder-sk commented on src/app/3d/qgs3deditutils.h:
```diff
@@ -0,0 +1,81 @@
+/***************************************************************************
+    qgs3deditutils.h
```
After seeing the whole code, I wonder whether it makes sense to actually move code from Qgs3DMapToolPointCloudChangeAttribute to Qgs3DEditUtils - because if Qgs3DMapToolPointCloudChangeAttribute is just a based class for concrete implementations, it is pretty much empty - and the code in Qgs3DEditUtils could simply stay there... (I don't think it would be needed in other parts of QGIS code anytime soon)

wonder-sk reviewed (commented):
Good stuff :tada:   I have left a bunch of comments, but we're on the right track!

Withalion committed:
Fix review issues
